Transaction, overview UI & code cleanup

oshorefueled opened the pull request:
This PR resolves issue #344. 

It also:
* contains code clean ups which include the addition of components.go, util.go and depricated.go file. Appropriate
   comments were added to each file describing the code that they contain. 
* modifies the progress bar widget. The default progress bar sets the max height to 4DP which doesn't allow a 8DP 
  height as specified in the overview mockup.
* adds a separator widget eliminating the need to initialize a Line object when initializing a page. And also repeatedly
  setting the width of the line. The same code is also present in PR #330.
* adds a Container type which is simply a wrapper to the layout.Inset type to help differentiate when it is used as 
   padding or margin. Its purpose is to help visualize the structure of a layout from the code.
* makes the transaction row layout reusable by both the transactions and overview page.

ademuanthony commented on cli/commands/show-transaction.go:
```diff
@@ -71,6 +72,30 @@ func (showTxCommand ShowTransactionCommand) Run(ctx context.Context, wallet wall
+		fmt.Println()
+		prompt := fmt.Sprintf("Enter (h)istory table, or (q)uit")
+
+		validateUserInput := func(userInput string) error {
```
```
validateUserInput := func(userInput string) error {
     if strings.EqualFold(userInput, "q") || strings.EqualFold(userInput, "h"){
        return nil
    }
    return fmt.Errorf("invalid response, try again")
}
```

Sirmorrison replied:
`  return fmt.Errorf("invalid response, try again")`

line 87 takes care of this.

ademuanthony commented on nuklear/pagehandlers/transaction_details.go:
```diff
@@ -3,11 +3,11 @@ package pagehandlers
 import (
 	"strconv"
 
+	"fmt"
```
re-arrange import

ademuanthony commented on cli/commands/show-transaction.go:
```diff
@@ -71,6 +72,30 @@ func (showTxCommand ShowTransactionCommand) Run(ctx context.Context, wallet wall
+			return nil
+		}
+
+		err = HistoryCommand{}.Run( ctx, wallet)
```
let us pass the `txHistoryOffset` here and update the `Run` so that the history page will return to the previous state

ademuanthony commented on cli/commands/history.go:
```diff
@@ -100,8 +122,15 @@ func (h HistoryCommand) Run(ctx context.Context, wallet walletcore.Wallet) error
 		if strings.EqualFold(userChoice, "q") {
 			break
-		} else if strings.EqualFold(userChoice, "m") {
+		} else if strings.EqualFold(userChoice, "n") {
```
update the comment on line 137, change the `m` to `n`

ademuanthony commented on cli/commands/history.go:
```diff
@@ -38,6 +38,8 @@ func (h HistoryCommand) Run(ctx context.Context, wallet walletcore.Wallet) error
 	var txHistoryOffset = 0
```
A `txHistoryOffset` field should be added to the `HistoryCommand` struct in line 17 and `ShowTransactionCommandArgs` on line 20 of `show-transaction.go` so that the user can be returned to the previous state of the history page before navigating to the details page.
Line 38 should now use `h.txHistoryOffset` and this value should be passed in `ShowTransactionCommandArgs` on line 142

raedah commented:
These screenshots dont look good with the hash wrapping.

ademuanthony commented on cli/commands/history.go:
```diff
@@ -16,6 +16,8 @@ import (
 // HistoryCommand enables the user view their transaction history.
 type HistoryCommand struct {
 	commanderStub
+	TxHistoryOffset   int
```
`int32`

ademuanthony commented on cli/commands/history.go:
```diff
@@ -35,58 +37,78 @@ func (h HistoryCommand) Run(ctx context.Context, wallet walletcore.Wallet) error
 	// ask user if to show next page, previous page, tx details or exit the loop
 	for {
-		transactions, err := wallet.TransactionHistory(int32(txHistoryOffset), txPerPage, nil)
+		transactions, err := wallet.TransactionHistory(int32(h.TxHistoryOffset), txPerPage, nil)
```
Use `int32` on line 19 so that you won't have to do type conversion here

ademuanthony commented on cli/commands/history.go:
```diff
@@ -100,17 +122,29 @@ func (h HistoryCommand) Run(ctx context.Context, wallet walletcore.Wallet) error
+		} else if strings.EqualFold(userChoice, "p") {
+			previous = true
 			fmt.Println() // print empty line before listing txs for next page
+			h.TxHistoryOffset -= int(txPerPage)
```
Use `int32 ` on line 19 so that you won't have to do type conversion here

ademuanthony commented on cli/commands/show-transaction.go:
```diff
@@ -17,7 +18,9 @@ type ShowTransactionCommand struct {
 type ShowTransactionCommandArgs struct {
-	TxHash string `positional-arg-name:"transaction hash" required:"yes"`
+	TxHash            string `positional-arg-name:"transaction hash" required:"yes"`
+	TxHistoryOffset   int
```
`int32`

itswisdomagain commented on cli/commands/history.go:
```diff
@@ -35,58 +37,78 @@ func (h HistoryCommand) Run(ctx context.Context, wallet walletcore.Wallet) error
-	var displayedTxHashes []string
 	var txPerPage int32 = walletcore.TransactionHistoryCountPerPage
+	var previous bool
+	var lastTX int
```
- `lastTx` instead of `lastTX`
- tx is not an int, do you need to rename this variable?

itswisdomagain commented on cli/commands/history.go:
```diff
@@ -16,6 +16,8 @@ import (
 type HistoryCommand struct {
 	commanderStub
+	TxHistoryOffset   int32
+	DisplayedTxHashes []string
```
These properties do not need to be public.

itswisdomagain commented on cli/commands/history.go:
```diff
@@ -35,58 +37,78 @@ func (h HistoryCommand) Run(ctx context.Context, wallet walletcore.Wallet) error
 	// ask user if to show next page, previous page, tx details or exit the loop
 	for {
-		transactions, err := wallet.TransactionHistory(int32(txHistoryOffset), txPerPage, nil)
+		transactions, err := wallet.TransactionHistory(h.TxHistoryOffset, txPerPage, nil)
```
Rename `h` to `history` please.

itswisdomagain commented on cli/commands/history.go:
```diff
@@ -16,77 +16,97 @@ import (
+			pageTxRows[i] = append(pageTxRows[i], tx.Type)
 		}
+
+		lastTx = len(transactions)
```
Do you mean `fetchedTxCount` or `totalFetchedTxs`? `lastTx` should be a `tx` (type `walletcore.Transaction`), not a number (type `int`).


UPDATE: After seeing more of your code, I think `previousPageTxCount` is more appropriate?

itswisdomagain commented on cli/commands/show-transaction.go:
```diff
@@ -71,6 +74,39 @@ func (showTxCommand ShowTransactionCommand) Run(ctx context.Context, wallet wall
 		}
 		termio.PrintStringResult(strings.TrimRight(detailedOutput.String(), " \n\r"))
+		fmt.Println()
+		prompt := fmt.Sprintf("Enter (h)istory table, or (q)uit")
```
Do we need to show this prompt if a user gets here by running `godcr show-transaction abcdefghash`?
If we don't, then you can check if `len(showTxCommand.displayedTxHashes) > 0` to know when a user gets here from the history command.

itswisdomagain commented on cli/commands/show-transaction.go:
```diff
@@ -17,7 +18,9 @@ type ShowTransactionCommand struct {
-	TxHash string `positional-arg-name:"transaction hash" required:"yes"`
+	TxHash            string `positional-arg-name:"transaction hash" required:"yes"`
+	txHistoryOffset   int32
+	displayedTxHashes []string
```
Add `txHistoryOffset` and `displayedTxHashes` to the `ShowTransactionCommand` struct, they are not command arguments.

Sirmorrison commented:
> 
> 
> First page is asking to enter `p` to show previous page:
> <img alt="Screenshot 2019-05-06 at 8 42 57 PM" width="598" src="https://user-images.githubusercontent.com/18400051/57250565-a7160700-703f-11e9-8f2e-25c1ab8d1b21.png">
> 
> Too much space after hitting quit:
> <img alt="Screenshot 2019-05-06 at 8 46 36 PM" width="512" src="https://user-images.githubusercontent.com/18400051/57250779-24417c00-7040-11e9-9866-5e8025b60fa4.png">

im not sure the code you re testing but this is my out put
![image](https://user-images.githubusercontent.com/27733432/57252073-3bce3400-7043-11e9-9c85-55b867a0ff73.png)

itswisdomagain commented:
@Sirmorrison, I get the previous page prompt on the first page when there's just one page of data to load. Temporarily increase your `txPerPage` number so that your app loads all txs in page 1 - you'll get the issue.

UMichael commented:
![image](https://user-images.githubusercontent.com/31141573/57325557-2f5ce080-7102-11e9-95d0-110cf3956202.png)
text overlaps on large string

Sirmorrison commented:
> 
> 
> ![image](https://user-images.githubusercontent.com/31141573/57325557-2f5ce080-7102-11e9-95d0-110cf3956202.png)
> text overlaps on large string in amount column

fixed please test and give feed back please

oshorefueled committed:
add comma to wallet balance

oshorefueled committed:
make colour and dimension corrections

- Correct inaccurate colours and dimension on the top bar. The colour used as background and separator colour was wrong. Receive and Send icons are also replaced with the samesize as those on the mockups.
- Add a container wrapper to differentiate when an inset is used as a margin or padding

oshorefueled committed:
correct dimensions and layout of transaction list

- Add a filtering method to filter transaction by direction. It replaces previous transactions filtering where transaction rows were being
  hidden.
- Correct dimensions and alignment of the transaction row

oshorefueled committed:
clean up no transactions layout

oshorefueled committed:
multi: add separator widget

- add separator widget and replace repetitive line implementation on pages
- move code that will be removed when all pages have been updated to the mockup designs to a deprecated file.
- move general page components such as the side nav and top bar to a
  component.go file.

oshorefueled committed:
multi: correct ui dimension on overview page

- modify progress bar, adding an exported Height field. The
  default progress bar sets the max height to 4DP which doesn't
  allow a 8DP height as specified in the overview mockup.

- make appropriate dimesion and text size corrections to the
  overview page.

- clean up code by adding code that aren't layout code to the util.go
  file.

oshorefueled committed:
ui: correct text sizes on non-decimal balances

- add comments to compontents, util and deprecated.go files
- remove top padding on transactions list

oshorefueled committed:
remove unused parameter

oshorefueled committed:
multi: add radius property to progressbar

- change color name from IconColor to Gray3

oshorefueled committed:
multi: make deep blue the default text colour

- add inactive colour to side nav text

JustinBeBoy commented on ui/decredmaterial/progressbar.go:
```diff
@@ -3,13 +3,97 @@
+		}),
+		layout.Stacked(func(gtx layout.Context) layout.Dimensions {
+			fillWidth := progressBarWidth * clamp1(p.Progress)
+			fillColor := p.Color
```
The progress color is "41BE53" and background is E6EAED

oshorefueled replied:
Yes, the colours for the track and progress are set on the pages using the progress bar widget

oshorefueled committed:
ui: reuse row layout on overview and about page

oshorefueled committed:
ui: correct padding for nav when expanded

oshorefueled committed:
ui: use icon size to a scale of 1 on privacy page
